go/consensus/tendermint/apps/staking/signing_rewards.go

```diff
@@ -0,0 +1,84 @@
+package staking
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"sort"
+
+	"github.com/oasislabs/oasis-core/go/common/crypto/signature"
+	"github.com/oasislabs/oasis-core/go/consensus/tendermint/abci"
+	stakingState "github.com/oasislabs/oasis-core/go/consensus/tendermint/apps/staking/state"
+	epochtime "github.com/oasislabs/oasis-core/go/epochtime/api"
+	staking "github.com/oasislabs/oasis-core/go/staking/api"
+)
+
+const (
+	SigningThresholdNumerator   = 3
+	SigningThresholdDenominator = 4
+)
+
+func (app *stakingApplication) updateEpochSigning(ctx *abci.Context, signingEntities []signature.PublicKey) error {
+	stakeState := stakingState.NewMutableState(ctx.State())
+
+	epochSigning, err := stakeState.EpochSigning()
+	if err != nil {
+		return fmt.Errorf("loading epoch signing info: %w", err)
+	}
+
+	oldTotal := epochSigning.Total
+	epochSigning.Total = oldTotal + 1
+	if epochSigning.Total <= oldTotal {
+		return fmt.Errorf("incrementing total blocks count: overflow, old_total=%d", oldTotal)
+	}
+
+	for _, entityID := range signingEntities {
+		oldCount := epochSigning.ByEntity[entityID]
+		epochSigning.ByEntity[entityID] = oldCount + 1
+		if epochSigning.ByEntity[entityID] <= oldCount {
+			return fmt.Errorf("incrementing count for entity %s: overflow, old_count=%d", entityID, oldCount)
+		}
+	}
+
+	stakeState.SetEpochSigning(epochSigning)
+
+	return nil
+}
+
+func (app *stakingApplication) rewardEpochSigning(ctx *abci.Context, time epochtime.EpochTime) error {
+	stakeState := stakingState.NewMutableState(ctx.State())
+
+	epochSigning, err := stakeState.EpochSigning()
+	if err != nil {
+		return fmt.Errorf("loading epoch signing info: %w", err)
+	}
+
+	stakeState.ClearEpochSigning()
+
+	if epochSigning.Total == 0 {
+		return nil
+	}
+
+	var eligibleEntities []signature.PublicKey
+	if epochSigning.Total > math.MaxUint64/SigningThresholdNumerator {
+		return fmt.Errorf("")
+	}
+	for entityID, count := range epochSigning.ByEntity {
+		if count > math.MaxUint64/SigningThresholdDenominator {
+			return fmt.Errorf("determining eligibility for entity %s: overflow in threshold comparison, count=%d", entityID, count)
+		}
+		if count*SigningThresholdDenominator < epochSigning.Total*SigningThresholdNumerator {
```

Comment:
`epochSigning.Total*SigningThresholdNumerator` could be computed just once outside the loop?

Reply:
yeah, although I leave that optimization to the compiler

Reply:
https://godbolt.org/z/CXMYMy
aw it doesn't

Reply:
eh I might as well, since you went through the trouble of commenting about it